File: 863. All Nodes Distance K in Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def distanceK(self, root, target, K):
        """
        :type root: TreeNode
        :type target: TreeNode
        :type K: int
        :rtype: List[int]
        First we make the directed graph into a undirected graph,
        and then we use bfs for finding distance of K
        """
        graph = collections.defaultdict(list)
        queue = [root]
        while queue:
            node = queue.pop(0)
            if node.left:
                graph[node].append(node.left)
                graph[node.left].append(node)
                queue.append(node.left)
            if node.right:
                graph[node].append(node.right)
                graph[node.right].append(node)
                queue.append(node.right)
        
        layer = [target]
        seen = {target}
        while K > 0:
            newlayer = []
            for node in layer:
                for newnode in graph[node]:
                    if newnode not in seen:
                        newlayer.append(newnode)
                        seen.add(newnode)
            K -= 1
            layer = newlayer
        return [node.val for node in layer]

File: 863. All Nodes Distance K in Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
/**
 * Convert into undirected graph, and bfs
 * Time O(n) build graph is one pass, and bfs is another pass
 * Space O(n)
 * Or we can only map the node with its parent and do a bfs
 * this will be a little bit faster
 */
class Solution {
    HashMap<Integer, HashSet<Integer>> graph;
    public List<Integer> distanceK(TreeNode root, TreeNode target, int K) {
        graph = new HashMap<>();
        buildGraph(root);
        HashSet seen = new HashSet<>();
        Queue<Integer> q = new LinkedList<>();
        q.add(target.val); seen.add(target.val);
        while (K > 0) {
            int size = q.size();
            for (int i=0; i<size; i++) {
                int val = q.poll();
                HashSet<Integer> set = graph.getOrDefault(val, new HashSet<>());
                for (int item : set) {
                    if (!seen.contains(item)) {
                        q.add(item);
                    }
                    seen.add(item);
                }
            }
            K--;
        }
        return new LinkedList<>(q);
    }
    private void buildGraph(TreeNode root) {
        if (root != null) {
            HashSet set;
            if (root.left != null) {
                set = graph.getOrDefault(root.val, new HashSet<>());
                set.add(root.left.val);
                graph.put(root.val, set);
                
                set = graph.getOrDefault(root.left.val, new HashSet<>());
                set.add(root.val);
                graph.put(root.left.val, set);
            }
            if (root.right != null) {
                set = graph.getOrDefault(root.val, new HashSet<>());
                set.add(root.right.val);
                graph.put(root.val, set);
                
                set = graph.getOrDefault(root.right.val, new HashSet<>());
                set.add(root.val);
                graph.put(root.right.val, set);
            }
            buildGraph(root.left);
            buildGraph(root.right);
        }
    }
}